---
title: "Analyzing single-cell bisulfite sequencing data with vmrseq"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Analyzing single-cell bisulfite sequencing data with vmrseq}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
devtools::load_all()
```

# Installation

This package will be submitted to Bioconductor. For now, you can install the development version from our [GitHub repository](https://github.com/nshen7/vmrseq) through running this:

```{r, eval = FALSE}
# install.packages("devtools")
devtools::install_github("nshen7/vmrseq")
```

After installation, load the package:

```{r setup, eval = FALSE}
library(vmrseq)
```

Note that you might also have to manually install one of the dependencies of vmrseq, the `bumphunter` package (using `BiocManager::install("bumphunter")`). 

# How to get help for vmrseq

Feel free to raise questions by opening issue(s) in the [vmrseq GitHub repository](https://github.com/nshen7/vmrseq)

# Background


# Analyze single-cell BS-seq data

## Input data

vmrseq takes processed and filtered single-cell bisulfite sequencing *binary* methylation values as input. After processing and filtering, each sequenced CpG takes a value of methylated or unmethylated in each cell.

### Process and filter cell read files
We have implemented a `data.pool` function to process individual-cell read files into a format that is suitable to input to the vmrseq framework. Here is an example of how to use this function:

```{r}
cell_1_path <- system.file("data", "cell_1.csv", package = "vmrseq")
cell_2_path <- system.file("data", "cell_2.csv", package = "vmrseq")
cell_3_path <- system.file("data", "cell_3.csv", package = "vmrseq")
```

```{r}
cell_list <- c(cell_1_path, cell_2_path, cell_3_path)
# data.pool()
```


SummarizedExperiment object (with NA-dropped sparseMatrix representation) using  function


### Apply vmrseq framework
In this vignette, we use a formatted example dataset built in the package called `se0`: 
```{r}
data(se0)
```

It is a `SummarizedExperiment` object that stores CpG sites as rows and cells as columns:
```{r}
GenomicRanges::granges(se0)
```

It has one `assay` slot called `M_mat` that contains the binary methylation values of individual cells at CpG sites:
```{r}
assays(se0)
```


Moreover, 
```{r}
assays(se0)$M_mat[5:10, 5:10]
```


```{r}
# gr <- vmrseq.smooth(se0)
# output <- vmrseq.fit(gr)
```

